test/TestUtilities/Test.Utility/ReadTimeoutHonoringSlowStream.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Test.Utility
+{
+    public class ReadTimeoutHonoringSlowStream : SlowStream
+    {
+        private readonly Stream _innerStream;
+        private readonly CancellationToken _cancellationToken;
+        private int _readTimeout = int.MaxValue;
+
+        public ReadTimeoutHonoringSlowStream(Stream innerStream)
+            : this(innerStream, CancellationToken.None)
+        {
+        }
+
+        public ReadTimeoutHonoringSlowStream(Stream innerStream, CancellationToken cancellationToken)
+            : base(innerStream, cancellationToken)
+        {
+            _innerStream = innerStream;
+            _cancellationToken = cancellationToken;
+        } 
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            OnRead?.Invoke(buffer, offset, count);
+            var read = _innerStream.Read(buffer, offset, count);
+
+            try
+            {
+                var expectedDelayInMs = DelayPerByte.TotalMilliseconds * read;
+                if (_readTimeout > expectedDelayInMs)
+                {
+                    Task.Delay(new TimeSpan(DelayPerByte.Ticks * read)).Wait(_cancellationToken);
+                }
+                else
+                {
+                    Task.Delay(_readTimeout).Wait(_cancellationToken);
+                    throw new IOException($"..timed out because no data was received for {_readTimeout}ms.", new TimeoutException());
+                }
+            }
+            catch (OperationCanceledException)
+            {
+            }
+
+            return read;
+        }
+
+        public override int ReadTimeout { get => _readTimeout; set => _readTimeout = value; }
```

Comment:
Is it necessary to create a backing field? Can't an auto-property be used?

Reply:
🙃 Fixed.